core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,710 @@
+            for it to be correct prediction.
+        list(int). IDs of training_data points which are misclassified by KNN.
+    """
+    training_data = add_top_similars(training_data, top_neighbours)
+
+    # No. of times a class has to appear in nearest neighbours of a program
+    # so that prediction is correct.
+    occurrence = 0
+
+    # Keep record of misclassified programs' ID.
+    misclassified_points = []
+
+    for pid in training_data:
+        similars = training_data[pid]['top_similar']
+        if pid in similars:
+            similars.remove(pid)
+        nearest_classes = [training_data[i]['class'] for (i, _) in similars]
+        cnt = collections.Counter(nearest_classes)
+        common = cnt.most_common(1)
+        training_data[pid]['prediction'] = common[0][0]
+
+        if training_data[pid]['prediction'] == training_data[pid]['class']:
+            # If prediction is correct then add no. of times the correct
+            # class has to appear in nearest neighbours to occurrence.
+            occurrence += common[0][1]
+        else:
+            # Else mark the ID fo program so that it will be
+            # trained using SVM later.
+            misclassified_points.append(pid)
+
+    # Calculate average of occurrence. This is used during prediction so
+    # that if winner of nearest neighbours appears less than the occurrence
+    # we consider that KNN has failed in prediction.
+    occurrence /= float(len(training_data))
+
+    return occurrence, misclassified_points
+
+
+class CodeClassifier(base.BaseClassifier):
+    """A class for code classifier that uses supervised learning to match
+    Python programs to an answer group. The classifier trains on programs
+    that exploration editors have assigned to an answer group.
+    """
+    def __init__(self):
+        super(CodeClassifier, self).__init__()
+        self.training_data = None
+        self.clf = None
+        self.token_to_id = None
+        self.T = None
+        self.K = None
+        self.occurrence = None
+        self.top = None
+        self.count_vector = None
+
+    def to_dict(self):
+        """Returns a dict representing this classifier.
+
+        Returns:
+            dict. A dictionary representation of classifier referred as
+            'classifier_data'. This data is used for prediction.
+        """
+        fingerprint_data = {
+            i: {
+                'fingerprint': self.training_data[i]['fingerprint'],
+                'class': self.training_data[i]['class']
+            } for i in self.training_data
+        }
+
+        classifier_data = {
+            'KNN': {
+                'token_to_id': self.token_to_id,
+                'T': self.T,
+                'K': self.K,
+                'top': self.top,
+                'occurrence': self.occurrence,
+                'fingerprint_data': fingerprint_data
+            },
+            'SVM': self.clf.__dict__,
```

Comment:
Yep, this looks good to me. But, I would think about doing some kind of validation instead of storing the params blindly. The dict should contain only the parameters that we have defined.